fhirr4/ballerina/src/main/resources/fhirservice/http_service_builder.bal

```diff
@@ -141,3 +147,66 @@ isolated function getSplitPath(string restPath) returns string[] {
     string[] requestPathParts = regex:split(path, "/").slice(1);
     return requestPathParts;
 }
+
+isolated function addPagination(r4:FHIRContext fhirContext, r4:Bundle bundle, int pageSize, string path) returns r4:Bundle {
+    r4:Bundle bundleWithPagination = bundle;
+    r4:BundleLink[] alllinks = [];
+    
+    // get current page
+    int currentpage = 1;
+    r4:RequestSearchParameter[]? pageParam = fhirContext.getRequestSearchParameter(OFFSET);
+    if pageParam is r4:RequestSearchParameter[] {
+        r4:RequestSearchParameter offsetParam = pageParam[0];
+        int|error offset = int:fromString(offsetParam.value);
+        if offset is int {
+            currentpage = (offset / pageSize) + 1;
+        } else {
+            // ignore error since we set the value to fhir context after validation
+        }
+    }
+
+    // populate self link
+    string selfUrl = path.includes(PAGE_QUERY_PARAM) ? path 
+                        : (path.includes("?") ? string `${path}&${PAGE_QUERY_PARAM}=${currentpage}` : string `${path}?${PAGE_QUERY_PARAM}=${currentpage}`);
+    r4:BundleLink selfLink = {
+        relation: "self",
+        url: string `${selfUrl}`
+    };
+
+    alllinks.push(selfLink);
+
+    // populate next link
+    int nextPage = currentpage + 1;
+    string nextUrl = getReplacedUrl(path, currentpage, nextPage);
+    r4:BundleLink nextLink = {
+        relation: "next",
+        url: string `${nextUrl}`
+    };
+    alllinks.push(nextLink);
+
+    if currentpage > 1 {
+        // previous link exists
+        // populate previous link
+        int prevPage = currentpage - 1;
+        string prevUrl = getReplacedUrl(path, currentpage, prevPage);
+        r4:BundleLink prevLink = {
+            relation: "prev",
+            url: string `${prevUrl}`
+        };
+        alllinks.push(prevLink);
+    }
+    bundleWithPagination.link = alllinks;
+    return bundleWithPagination;
+}
+
+isolated function getReplacedUrl(string path, int current, int replacement) returns string {
+    string replacedUrl = "";
+    if path.includes(PAGE_QUERY_PARAM) {
+        replacedUrl = regex:replace(path, string `page=${current}`, string `page=${replacement}`);
```

Comment:
regex evaluation is somewhat expensive. shall we have a util for build this url? it will accept the page value and it will template the page search param in the for the given url.

Reply:
added